src/rpc/server.cpp

```diff
@@ -490,7 +491,42 @@ UniValue CRPCTable::execute(const std::string &strMethod, const UniValue &params
     try
     {
         // Execute
-        return pcmd->actor(params, false);
+        auto paramRange = rpcCvtTable.find(strMethod);
+        if (paramRange != rpcCvtTable.end()) {
+            auto numRequired = paramRange->second.first.size();
+            auto numOptional = paramRange->second.second.size();
+            if (params.size() < numRequired || numRequired + numOptional < params.size()) {
+                std::string helpMsg;
+                try {
+                    // help gets thrown – if it doesn’t throw, then no help message
+                    pcmd->actor(params, true);
+                } catch (const std::runtime_error& err) {
+                    helpMsg = std::string("\n\n") + err.what();
+                }
+                throw JSONRPCError(
+                    RPC_INVALID_PARAMS,
+                    strprintf(
+                            "%s for method `%s`. Needed %s, but received %u%s",
+                            params.size() < numRequired
+                            ? "Not enough parameters"
+                            : "Too many parameters",
+                            strMethod,
+                            numOptional == 0
+                            ? strprintf("exactly %u", numRequired)
+                            : strprintf("from %u to %u", numRequired, numRequired + numOptional),
```

Comment:
This should use the same language as in `FormatConversionFailure`. So if my above suggestion is accepted:
```suggestion
                            : strprintf("at least %u and at most %u", numRequired, numRequired + numOptional),
```